src/cascadia/WindowsTerminal/WindowEmperor.cpp

```diff
@@ -18,600 +30,1024 @@ using namespace ::Microsoft::Console;
+        __assume(false);
     }
 
-    // At this point, there weren't any pending refrigerated threads we could
-    // just use. That's fine. Let's just go create a new one.
+    _app = winrt::TerminalApp::App{};
+    _app.Logic().ReloadSettings();
 
-    window = std::make_shared<WindowThread>(_app.Logic(), args, _manager, peasant);
+    _createMessageWindow(windowClassName.c_str());
+    _setupGlobalHotkeys();
+    _checkWindowsForNotificationIcon();
 
-    std::weak_ptr<WindowEmperor> weakThis{ weak_from_this() };
+    // When the settings change, we'll want to update our global hotkeys
+    // and our notification icon based on the new settings.
+    _app.Logic().SettingsChanged([this](auto&&, const TerminalApp::SettingsLoadEventArgs& args) {
+        if (SUCCEEDED(args.Result()))
+        {
+            _assertIsMainThread();
+            _setupGlobalHotkeys();
+            _checkWindowsForNotificationIcon();
+        }
+    });
 
-    // Increment our count of window instances _now_, immediately. We're
-    // starting a window now, we shouldn't exit (due to having 0 windows) till
-    // this window has a chance to actually start.
-    // * We can't just put the window immediately into _windows right now,
-    //   because there are multiple async places where we iterate over all
-    //   _windows assuming that they have initialized and we can call methods
-    //   that might hit the TerminalPage.
-    // * If we don't somehow track this window now, before it has been actually
-    //   started, there's a possible race. As an example, it would be possible
-    //   to drag a tab out of the single window, which would create a new
-    //   window, but have the original window exit before the new window has
-    //   started, causing the app to exit.
-    // Hence: increment the number of total windows now.
-    _windowThreadInstances.fetch_add(1, std::memory_order_relaxed);
+    {
+        const wil::unique_environstrings_ptr envMem{ GetEnvironmentStringsW() };
+        const auto env = stringFromDoubleNullTerminated(envMem.get());
+        const auto cwd = wil::GetCurrentDirectoryW<std::wstring>();
+        const auto showCmd = gsl::narrow_cast<uint32_t>(nCmdShow);
 
-    std::thread t([weakThis, window]() {
-        try
+        // Restore persisted windows.
+        const auto state = ApplicationState::SharedInstance();
+        const auto layouts = state.PersistedWindowLayouts();
+        if (layouts && layouts.Size() > 0)
         {
-            window->CreateHost();
+            _needsPersistenceCleanup = true;
 
-            if (auto self{ weakThis.lock() })
+            uint32_t startIdx = 0;
+            for (const auto layout : layouts)
             {
-                self->_windowStartedHandlerPostXAML(window);
+                hstring args[] = { L"wt", L"-w", L"new", L"-s", winrt::to_hstring(startIdx) };
+                _dispatchCommandline({ args, cwd, showCmd, std::move(env) });

```

Comment:
won't this move from `env` multiple times?

Reply:
The reason this worked is because `env` is `const` and so it doesn't move it, it makes a copy. 🥴